compiler/src/dotty/tools/dotc/semanticdb/TypeOps.scala

```diff
@@ -0,0 +1,304 @@
+package dotty.tools
+package dotc
+package semanticdb
+
+import core.Symbols._
+import core.Contexts.Context
+import core.Types._
+import core.Annotations.Annotation
+import core.Flags
+import core.Names.Name
+import core.StdNames.tpnme
+import ast.tpd._
+
+import collection.mutable
+
+import dotty.tools.dotc.{semanticdb => s}
+
+class TypeOps:
+  import SymbolScopeOps._
+  private val paramRefSymtab = mutable.Map[(LambdaType, Name), Symbol]()
+  private val refinementSymtab = mutable.Map[(RefinedType, Name), Symbol]()
+  given typeOps: TypeOps = this
+  extension (tpe: Type)
+    def toSemanticSig(using LinkMode, Context, SemanticSymbolBuilder)(sym: Symbol): s.Signature =
+      def enterParamRef(tpe: Type): Unit =
+        tpe match {
+          case lam: LambdaType =>
+            paramRefSymtab((lam, sym.name)) = sym
+
+          // for class constructor
+          // class C[T] { ... }
+          case cls: ClassInfo if sym.info.isInstanceOf[LambdaType] =>
+            val lam = sym.info.asInstanceOf[LambdaType]
+            cls.cls.typeParams.foreach { param =>
+              paramRefSymtab((lam, param.name)) = param
+            }
+
+          // type X[T] = ...
+          case tb: TypeBounds =>
+            enterParamRef(tb.lo)
+            enterParamRef(tb.hi)
+
+          case _ => ()
+        }
+
+      def enterRefined(tpe: Type): Unit =
+        tpe match {
+          case refined: RefinedType =>
+            val key = (refined, sym.name)
+            refinementSymtab(key) = sym
+
+          case rec: RecType =>
+            enterRefined(rec.parent)
+
+          case cls: ClassInfo
+            if (cls.cls.name == tpnme.REFINE_CLASS) =>
+            sym.owner.owner.info match {
+              // def foo(x: Parent { refinement }) = ...
+              case refined: RefinedType =>
+                val key = (refined, sym.name)
+                refinementSymtab(key) = sym
+
+              // type x = Person { refinement }
+              case tb: TypeBounds =>
+                // tb = TypeBounds(
+                //   lo = RefinedType(...)
+                //   hi = RefinedType(...)
+                // )
+                enterRefined(tb.lo)
+                enterRefined(tb.hi)
+
+              // def s(x: Int): { refinement } = ...
+              case expr: ExprType =>
+                enterRefined(expr.resType)
+              case m: MethodType =>
+                enterRefined(m.resType)
+              case _ => ()
+            }
+          case _ => ()
+        }
+      enterParamRef(sym.owner.info)
+      enterRefined(sym.owner.info)
+
+      def loop(tpe: Type): s.Signature = tpe match {
+        case mt: MethodType =>
+          val stparams = Some(s.Scope())
+          val paramss =
+            if (sym.rawParamss.nonEmpty) sym.rawParamss else sym.paramSymss
```

Comment:
`sym.paramSymss` has also the type parameters, not just the term parameters, so `stparams` should not be empty here.

Reply:
an extension method can also have type parameters after term parameters, so thats another thing to look out for

Reply:
Ah, that's right, fixed in https://github.com/lampepfl/dotty/pull/12885/commits/75bc739cbeae088a2a0604b924ce8c6eaeaa86cc by filtering out the type parameters here, and the type parameters are set in the case pattern for `PolyType`

Reply:
I would lift the type parameters to the front in this case. MethodInformation doesn't need to correspond to how the code is parsed.

Reply:
it gets more tricky when you start having type parameters that have a dependent type on a previous term parameter, so would forward references break anything?

Reply:
oh, I assumed that the type parameters always come before term parameters.
To handle this, I think I should change the way to collect parameters:
Currently, we're collecting parameters of MethodOrPoly by `paramSymss`, but we should gather the parameters as  other `TypeLambda`s does.

Reply:
And I think what we can do to this case is lifting the type parameters to the front because SemanticDB doesn't have the capability to express type parameters that depends on the first term parameters.
- SemanticDB's MethodSignature doesn't care about the order of type params/term params.
- SemanticDB can't express nested method signatures.

Reply:
Lifted the type parameter to "front" https://github.com/lampepfl/dotty/pull/12885/commits/a8a137c36977125b3d65e2c2289268961ce2db14